### ISSUE TITLE
相同的参数，由于随机过程导致的差异常大

### ISSUE THREAD

**Issue by iminders** [iminders/tbase#2]
1. 设置随机数种子，使模型训练可复现？
2. 由于撮合成交机制，微小的价格差异，导致行为差别非常大。微小的随机差异，可能使portfolio值相差非常大，容易形成非常剧烈的抖动现象， 如何避免这种情况？

这两种方法可行性探索，如果可行则更新到repo中 @iminders

**Comment by iminders** [comment]
https://github.com/iminders/tbase/pull/5

设置全局种子，并不能使实验可复现，随机过程并不一致

**Comment by iminders** [comment]
https://github.com/iminders/tbase/pull/10
使用一个dummy agent检查随机过程的可复现性

**Comment by iminders** [comment]
> #10
> 使用一个dummy agent检查随机过程的可复现性

不使用多进程时，可以复现模型的运行过程。

多进程场景下如何使复现模型？

**Comment by iminders** [comment]
https://pytorch.org/docs/stable/notes/randomness.html

Completely reproducible results are not guaranteed across PyTorch releases, individual commits or different platforms. Furthermore, results need not be reproducible between CPU and GPU executions, even when using identical seeds.
....
A number of operations have backwards that use atomicAdd, in particular torch.nn.functional.embedding_bag(), torch.nn.functional.ctc_loss() and many forms of pooling, padding, and sampling. There currently is no simple way of avoiding non-determinism in these functions.

**Comment by iminders** [comment]
不可复现的原因有以下两方面因素：
1. 中心OrnsteinUhlenbeckProcess, 由于跟步数相关，多线程/进程会导致随机过程与并行时的真实执行的顺序有关，多进程并行的场景下，不能完全控制执行顺序

2.[ pytorch内部op(如: sampling,  backwards ）在并行的场景下无法保证可复现](https://pytorch.org/docs/stable/notes/randomness.html)

**Comment by iminders** [comment]
#10 增加了非多进程运行方式， 可以完美复现结果
